src/charm.py

```diff
@@ -274,6 +277,27 @@ def _command(self) -> str:
 
         return " ".join(command)
 
+    def _on_validate_configuration(self, event: ActionEvent) -> None:
```

Comment:
Should we be using this during layer construction?

Reply:
It's not actually possible at least during initial layer construction (the container will not be running). When a new config is generated, we _could_, but the question then becomes what we would do with it.

Prometheus already goes into `BlockedStatus` if the service cannot start, which gives admins a chance to run this action.

If anything, validation in this sense should extract the relevant portions from `promtool` into `cos-tool` and validate rules/configs per-relation in a manner similar to alerts. I opened an issue to track this [here](https://github.com/canonical/prometheus-k8s-operator/issues/314).